src/transaction/log_manager.c

```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+#endif
+  if (cdc_Gl.queue_size >= MAX_CDC_LOGINFO_QUEUE_SIZE || cdc_logInfo_queue->is_full ())
+    {
+#if 1				//JOOHOK
+      _er_log_debug (ARG_FILE_LINE, "cdc_log_producer : queue size and number of entry is exceeded");
+#endif
+      if (cdc_Gl.is_finalize == true)
+	{
+	  pthread_mutex_unlock (&cdc_Gl.queue_consume_lock);
+	  goto end;
+	}
+      pthread_cond_wait (&cdc_Gl.queue_consume_cond, &cdc_Gl.queue_consume_lock);
+    }
+  pthread_mutex_unlock (&cdc_Gl.queue_consume_lock);
+
+  pthread_mutex_lock (&cdc_Gl.is_finalize_lock);
+#if 1				//JOOHOK
+  _er_log_debug (ARG_FILE_LINE, "cdc_log_producer : finalize");
+#endif
+  if (cdc_Gl.is_finalize == true)
+    {
+#if 1				//JOOHOK
+      _er_log_debug (ARG_FILE_LINE, "cdc_log_producer : signal finailize ");
+#endif
+      goto end;
+    }
+  pthread_mutex_unlock (&cdc_Gl.is_finalize_lock);
+
+  temp_logbuf[0].log_page_p = (LOG_PAGE *) PTR_ALIGN (temp_logbuf[0].log_page, MAX_ALIGNMENT);
+  temp_logbuf[0].pageid = -1;
+  temp_logbuf[1].log_page_p = (LOG_PAGE *) PTR_ALIGN (temp_logbuf[1].log_page, MAX_ALIGNMENT);
+  temp_logbuf[1].pageid = -1;
+
+  LSA_COPY (&cur_log_rec_lsa, &cdc_Gl.next_lsa);	/*lsa_to_process will be defined when get_lsa or get_log_item protocol is called */
+  LSA_COPY (&process_lsa, &cur_log_rec_lsa);
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  pthread_mutex_lock (&cdc_Gl.is_finalize_lock);
+
+  /*fetch log page */
+  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+    {
+      goto end;
+    }
+
+  tmpbuf_index = process_lsa.pageid % 2;
+  temp_logbuf[tmpbuf_index].pageid = process_lsa.pageid;
+  memcpy (temp_logbuf[tmpbuf_index].log_page_p, log_page_p, IO_MAX_PAGE_SIZE);
+
+
+  log_info_entry = (CDC_LOGINFO_ENTRY *) malloc (sizeof (CDC_LOGINFO_ENTRY));
+  if (log_info_entry == NULL)
+    {
+      goto end;
+    }
+
+  log_info_entry->length = 0;
+  LSA_SET_NULL (&log_info_entry->start_lsa);
+  log_info_entry->log_info = NULL;
+
+  if (!LSA_ISNULL (&process_lsa))
+    {
+#if 1				//JOOHOK
+      _er_log_debug (ARG_FILE_LINE, "cdc_log_producer : start produce : %lld | %d", LSA_AS_ARGS (&process_lsa));
+#endif
+      log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+
+      log_type = log_rec_header->type;
+      trid = log_rec_header->trid;
+      LSA_COPY (&next_log_rec_lsa, &log_rec_header->forw_lsa);
+
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_header), &process_lsa, log_page_p);
+
+      switch (log_type)
+	{
+	case LOG_COMMIT:
+	case LOG_ABORT:
+	  LOG_REC_DONETIME * donetime;
+
+	  LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*donetime), &process_lsa, log_page_p);
+	  if ((process_lsa.pageid % 2) != tmpbuf_index)
+	    {
+	      tmpbuf_index = (tmpbuf_index + 1) % 2;
+	      temp_logbuf[tmpbuf_index].pageid = process_lsa.pageid;
+	      memcpy (temp_logbuf[tmpbuf_index].log_page_p, log_page_p, IO_MAX_PAGE_SIZE);
+	    }
```

Comment:
How about introducing a new macro that covers this execution similar to the LOG_READ_ADVANCE_WHEN_DOESNT_FIT macro. It is nice if we can remove memcpy function call.